docs/randomized_testing.md

```diff
@@ -168,6 +168,535 @@ roughly the same things from the user (which operations to run, how to generate
+}
+
+func (w *Workload) CallA(ctx context.Context, x int) error {
+    if y, err := w.a.Get().A(ctx, x); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+
+func (*Workload) GenCallA(r *rand.Rand) int {
+    ...
+}
+
+func (w *Workload) CallB(ctx context.Context) error {
+    w.b.Get().B(ctx)
+    return nil
+}
+
+func TestWorkload(t *testing.T) {
+    s, err := sim.New[Workload](sim.Options{...})
+    if err != nil {
+        t.Fatal(err)
+    }
+    results, err := s.Simulate(10 * time.Second)
+    if err != nil {
+        t.Fatal(err)
+    }
+}
+```
+
+**Pros.**
+This API is arguably more ergonomic than the other options.
+
+**Cons.**
+This option is very magical. Rather than calling library functions like
+`sim.RegisterOp`, for example, you have to know the special syntax to create an
+op. Moreover, the `Fake` methods are only called once during initialization, so
+it's odd they are methods. It would make more sense if they were part of `Init`.
+
+### Option 4
+
+Option 4 is identical to Option 3, but rather than having `FakeFoo` methods, the
+`Init` method returns a slice of fakes using the existing `weavertest.Fake`
+function.
+
+```go
+func (w *Workload) Init(context.Context) ([]weavertest.FakeComponent, error) {
+    w.replies = map[int]int{}
+    fakes := []weavertest.FakeComponent{weavertest.Fake[B](&fakeb{})}
+    return fakes, nil
+}
+```
+
+The returned fakes can be optional, allowing for a simpler `Init` method for
+workloads without fakes:
+
+```go
+func (w *Workload) Init(context.Context) error {
+    // No fakes here.
+    w.replies = map[int]int{}
+    return fakes, nil
+}
+```
+
+**Pros.**
+This option re-uses the existing `weavertest` mechanism to register fakes. It
+also avoids the magical `Fake` functions.
+
+**Cons.**
+If `Init` is required to return `[]weavertest.FakeComponent`, then the `Init`
+signature is complicated even for workloads that don't need fakes. If it is
+optional, then there's complexity in understanding `Init` can have two different
+signatures. In contrast with Option 3, the simulator does not know ahead of time
+which components are faked. A workload could also theoretically return a
+different set of fakes every time `Init` is called, which is not intended.
+
+### Option 5
+
+Option 5 is identical to Option 3, except that fakes are fields in the workload
+struct. For example, in the code snippet below, the tag `` `fake:"b"` ``
+indicates that the `fakeb` field is a fake for the `weaver.Ref[B]` field named
+`b`. You can initialize fakes in the `Init` function.
+
+```go
+type Workload struct {
+    a weaver.Ref[A]
+    b weaver.Ref[B]
+    fakeb fakeb `fake:"b"`
+    replies map[int]int
+}
+
+func (w *Workload) Init(context.Context) error {
+    w.replies = map[int]int{}
+    // Initialize w.fakeb here.
+    return nil
+}
+```
+
+**Pros.**
+No magic `Fake` functions.
+
+**Cons.**
+Magic struct tags. Also, it is awkward to fake a component to which we don't
+need a direct reference.
```

Comment:
What if we change to something like:

```go
type Workload struct {
    a weaver.Ref[A]
    b weavertest.Fake[B,fakeb]
    replies map[int]int
}
```

Would Fake[I,F] be something applicable to weavertest as well, which might make it more palatable.

E.g., chain_test.go's fake could instead by supplied as follows:

```go
runner.Test(t, func(t *testing.T, a weavertest.Fake[A, fake]) {
    f := a.Fake()  // type *fake
    i := a.Component()  // type A
    ...
})
```

We can then get rid of weavertest.Fake and weavertest.FakeComponent. We should try writing some user documentation for this and see how it looks.

Reply:
Added this option to the doc. I can work on writing user docs.